paasta_tools/iptables.py

```diff
@@ -0,0 +1,145 @@
+
+import collections
+import re
+
+import iptc
+
+
+CHAIN_REGEX = re.compile('-N ([^ ]+)$')
+
+
+class Rule(collections.namedtuple('Rule', (
+    'protocol',
+    'src',
+    'dst',
+    'target',
+    'matches',
+))):
+    """Rule representation.
+
+    Working with iptc's rule classes directly doesn't work well, since rules
+    represent actual existing iptables rules, and changes are applied
+    immediately. They're also difficult to compare.
+    """
+
+    @classmethod
+    def from_iptc(cls, rule):
+        fields = {
+            'protocol': rule.protocol,
+            'src': rule.src,
+            'dst': rule.dst,
+            'target': rule.target.name,
+            'matches': (),
+        }
+
+        for match in rule.matches:
+            fields['matches'] += ((
+                match.name,
+                tuple((param, value) for param, value in match.parameters.items())
+            ),)
+
+        return cls(**fields)
+
+    def to_iptc(self):
+        rule = iptc.Rule()
+        rule.protocol = self.protocol
+        rule.src = self.src
+        rule.dst = self.dst
+        rule.create_target(self.target)
+        for name, params in self.matches:
+            match = rule.create_match(name)
+            for key, value in params:
+                setattr(match, key, value)
+        return rule
+
+
+class ChainDoesNotExist(Exception):
+    pass
+
+
+def all_chains():
+    return {chain.name for chain in iptc.Table(iptc.Table.FILTER).chains}
+
+
+def ensure_chain(chain, rules):
+    """Idempotently ensure a chain exists and has an exact set of rules.
+
+    This function creates or updates an existing chain to match the rules
+    passed in.
+
+    This function will not reorder existing rules, but any new rules are always
+    inserted at the front of the chain.
+    """
+    try:
+        current_rules = list_chain(chain)
+    except ChainDoesNotExist:
+        create_chain(chain)
+        current_rules = set()
+
+    for rule in rules:
+        if rule not in current_rules:
+            insert_rule(chain, rule)
+
+    # The insertion stage needs "rules" to be ordered since insertion order
+    # matters, but the deletion phase doesn't, so we turn it into a set to
+    # speed up this loop.
+    rules = set(rules)
+    for rule in current_rules:
+        if rule not in rules:
+            delete_rule(chain, rule)
+
+
+def ensure_rule(chain, rule):
+    rules = list_chain(chain)
+    if rule not in rules:
+        insert_rule(chain, rule)
+
+
+def insert_rule(chain_name, rule):
+    print('adding rule to {}: {}'.format(chain_name, rule))
+    chain = iptc.Chain(iptc.Table(iptc.Table.FILTER), chain_name)
+    chain.insert_rule(rule.to_iptc())
+
+
+def delete_rule(chain_name, rule):
+    print('deleting rule from {}: {}'.format(chain_name, rule))
+    chain = iptc.Chain(iptc.Table(iptc.Table.FILTER), chain_name)
+    for potential_rule in chain.rules:
+        if Rule.from_iptc(potential_rule) == rule:
```

Comment:
Yeah, I was using it here originally but temporarily deleted it. I'll need to figure out how to refactor this function to be fast anyway, just doing it without autocommit won't actually speed things up. Not using autocommit is more of a correctness thing, if I'm reading correctly their docs.